File: shared/src/main/java/com/screentime/shared/model/TimeRequest.kt
package com.screentime.shared.model

import java.time.Instant

data class TimeRequest(
    val id: String,
    val appPackage: String,
    val requestedMinutes: Int,
    val status: Status = Status.Pending,
    val approvedMinutes: Int? = null,
    val createdAt: Instant = Instant.now(),
    val respondedAt: Instant? = null,
) {
    enum class Status { Pending, Approved, Denied }

    /** True while an approved grant's window ([respondedAt] + granted minutes) hasn't elapsed yet. */
    fun isActiveGrant(now: Instant = Instant.now()): Boolean {
        val expiry = grantExpiresAt() ?: return false
        return now.isBefore(expiry)
    }

    fun grantExpiresAt(): Instant? {
        if (status != Status.Approved) return null
        val grantedMinutes = approvedMinutes ?: requestedMinutes
        return respondedAt?.plusSeconds(grantedMinutes * 60L)
    }
}

File: web/src/models/TimeRequest.ts
export type TimeRequestStatus = 'Pending' | 'Approved' | 'Denied';

export interface TimeRequest {
  id: string;
  appPackage: string;
  requestedMinutes: number;
  status: TimeRequestStatus;
  approvedMinutes: number | null;
  createdAt: Date;
  respondedAt: Date | null;
}

export function grantExpiresAt(request: TimeRequest): Date | null {
  if (request.status !== 'Approved') return null;
  const grantedMinutes = request.approvedMinutes ?? request.requestedMinutes;
  if (!request.respondedAt) return null;
  return new Date(request.respondedAt.getTime() + grantedMinutes * 60_000);
}

/** True while an approved grant's window (respondedAt + granted minutes) hasn't elapsed yet. */
export function isActiveGrant(request: TimeRequest, now: Date = new Date()): boolean {
  const expiry = grantExpiresAt(request);
  if (!expiry) return false;
  return now.getTime() < expiry.getTime();
}
